rclcpp/src/rclcpp/executors/executor_entities_collector.cpp

```diff
@@ -0,0 +1,286 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <set>
+
+#include "rclcpp/executors/executor_entities_collector.hpp"
+#include "rclcpp/executors/executor_notify_waitable.hpp"
+#include "rclcpp/node_interfaces/node_base_interface.hpp"
+
+namespace rclcpp
+{
+namespace executors
+{
+
+ExecutorEntitiesCollector::ExecutorEntitiesCollector(
+  std::function<void(void)> on_notify_waitable_callback)
+: notify_waitable_(std::make_shared<ExecutorNotifyWaitable>(on_notify_waitable_callback))
```

Comment:
In hindsight, it's actually better as it is right now for the events-executor because I should handle it separately.

Reply:
I am currently manually adding it as part of my `on_notify_waitable_callback` function.

Reply:
What do you think of having the executor own the notify waitable and pass it to the entities collector constructor?

Reply:
I'm good with that, I will make the change.

Reply:
pushed.  From here on, I will be extra sure to not rebase this branch since I know you are using it.  My feature branch for the actual executor implementation may be rebased, so don't depend on that one.